### DISCUSSION TITLE
Sand Table on OrangePi

### DISCUSSION THREAD

**Discussion by MaxAltera** [rdudhagra/Sand-Table#37] [Q&A] [answered]
Good afternoon.
First of all, I would like to thank you for your great work.
Due to the fact that Raspberry is very expensive for me, I decided to try using OrangePi Zero2 to install the controller, but at the time of installation, some difficulties began to arise.
Please tell me on which operating system (Symbian or Debian) can I install this program? I tried to install SandTable on OrangePi Zero2 on Armbian, but the system writes that the outdated version of node js (the latest version is 12.22). And tell me how you can make your own control programs? Maybe someone has already installed SandTable on OrangePi?

**Comment by rdudhagra** [comment, marked as answer]
Hi @MaxAltera,

I used Raspberry Pi OS so I'd recommend something Debian-based just to stay as consistent as possible. I've never used the OrangePi but I did encounter a NodeJS error on my Raspberry Pi when I was building my table. Did you try manually installing a later version of NodeJS? I have instructions for that in the [build guide](https://github.com/rdudhagra/Sand-Table/tree/master/Build%20Guide#raspberry-pi-setupwebserver-installation), but if you're still having errors feel free to paste the terminal output below and I can help diagnose. 

The build of Raspberry Pi OS I've been using is based off Debian 10. If you run `cat /etc/os-release` what output do you get? You might just be on a really old version of Debian.

**Reply by MaxAltera** [reply]
thanks for the reply. As I understand it, the light version is used here. I'll try to install debian 10

```bash
_Welcome to Orange Pi 3.0.6 Bullseye with Linux 5.16.17-sun50iw9

System load:   45%              Up time:       4 min
Memory usage:  14% of 984M      IP:            192.168.0.102
CPU temp:      55°C             Usage of /:    6% of 29G

[ 0 security updates available, 9 updates total: apt upgrade ]
Last check: 2023-02-07 16:20

New to Orange Pi? Support: http://www.orangepi.org

root@orangepizero2:~# Welcome to Orange Pi 3.0.6 Bullseye with Linux 5.16.17-sun50iw9

root@orangepizero2:~# su -
root@orangepizero2:~# cat /etc/issue
Orange Pi 3.0.6 Bullseye \l

root@orangepizero2:~# cat /etc/debian_version
11.4
root@orangepizero2:~#_
```
I still managed to install the desired version of mpm. to do this, in step 4, you need to enter the following commands:
```bash
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
export NVM_DIR="$HOME/.nvm"
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
nvm install 14.8.0
npm install -g npm@6.14.7
```
now I have a problem with point 5. when executing "npm install", I end up installing with an error:
![2023-02-07_22-22-54](https://user-images.githubusercontent.com/86003419/217346178-7f95322d-1b74-47e8-999c-fd59cc747211.png)
![2023-02-07_22-23-16](https://user-images.githubusercontent.com/86003419/217346183-e2080e00-a240-4b28-a7cb-a06fd1c24246.png)
![2023-02-07_22-23-44](https://user-images.githubusercontent.com/86003419/217346184-0e0175f1-6157-4ef7-af3e-e2cc7faa3d2f.png)

**Reply by rdudhagra** [reply]
`node-gyp` is always a huge pain to get to install on non-standard systems....

Try running `sudo apt install libvips-dev`? ([source](https://github.com/lovell/sharp/issues/1882#issuecomment-534241405))

**Reply by MaxAltera** [reply]
The suggested installation command does not work. I looked at the list of available libvips packages - there is no necessary package. I want to install https://github.com/libvips/libvips/releases/download/v8.14.1/vips-8.14.1.tar.xz , but while I'm looking for instructions on how to install applications from the archives.
![IMG_20230208_105422](https://user-images.githubusercontent.com/86003419/217491573-1b9e6a50-2cd4-410d-b350-511b6071c0b9.jpg)
![IMG_20230208_095502](https://user-images.githubusercontent.com/86003419/217491628-e81be864-d605-403b-8ece-af05b8625374.jpg)
![IMG_20230208_100458](https://user-images.githubusercontent.com/86003419/217491639-2d9daa7f-73a7-440a-b91a-b959404175c8.jpg)

**Reply by rdudhagra** [reply]
I'm honestly not sure how to fix this. From the logs, it looks like you have a version of `libvips` that's too old for the `sharp` library that `npm` is trying to install. Is there any way you can build/install the newer version of `libvips` that you linked above from source?

**Reply by MaxAltera** [reply]
Yes, I see that I have an older version, but apt does not offer me a newer one. I tried to build from a package on github to get a deb file, but it was unsuccessful. I collect information online. This is the main task now to move forward. Which version of linux and Debian are you using? I'll try to set myself the same

**Comment by MaxAltera** [comment]
I gave up the idea to launch OrangePi. Temporarily) But the idea of running a sand table haunts me. I bought a Raspberry Pi 3B+. Installed all packages according to the instructions. But in the end, on the website "Welcome to nginx!". I started over several times. The result is the same. Where could I have gone wrong?
![2023-02-28_21-04-15](https://user-images.githubusercontent.com/86003419/221942807-11d65867-2205-4fb2-bd1b-b88bf6c5b610.png)
![2023-02-28_21-04-43](https://user-images.githubusercontent.com/86003419/221942812-a1a119df-cbb3-445e-9b9c-09bc3aade2a3.png)

**Reply by rdudhagra** [reply]
Sorry for the super-late reply. Can you give me the output of `sudo systemctl status nginx`? And confirm that the `nginx.conf` is correctly set up/etc per the instructions?

**Reply by MaxAltera** [reply]
![2023-03-27_20-27-31](https://user-images.githubusercontent.com/86003419/228021400-9e0d7d82-2f8a-4b93-a95a-b51f1b6a371a.png)Thank you for your support. To be honest, I was already desperate to set up this project. After your answer, I once again carefully followed your instructions. My mistake was in the wrong nginx installation. When checking the directory "/etc/nginx", I understand that it is empty. Should it be empty? I may seem insistent, but maybe you have a web server image for Raspberry, so that stupid users like me can download it to an SD card and run it, and not build a project from source. I started to get to this stage by manually installing specific versions of packages from apt, otherwise some packages are incompatible with each other.

The question also arose, is it necessary to use Marlin to control engines, because there is a more affordable and compact analogue on GRBL?

**Reply by rdudhagra** [reply]
Yea your `nginx` install seems non-standard. Does `sudo systemctl status nginx` return anything? What path comes up when you run `which nginx`? You might have to uninstall and reinstall nginx using `apt` but not sure what your system looks like.

Try following this link to fix your install: https://stackoverflow.com/questions/12362967/how-can-i-restore-etc-nginx

I totally agree that a prebuilt image would be the best for this; although, I'm inexperienced in making images for distribution and didn't want to accidentally expose passwords/private keys.

**Reply by MaxAltera** [reply]
Once again, I reinstall all the packages according to your instructions, but the result is the same. Apparently, I constantly get up on the same rake) While I am studying the features of installing nginx.
A card image is a copy of a microSD card saved to a file. I use these as an emergency backup. It is convenient to do it with the Win32 Disk Imager program. In addition, SandTable is a standalone device - there is nothing to hack here. But it would greatly help to develop your creation) But this is only up to your desire. While I and my fellow programmers are trying to solve this problem to the end)

**Reply by rdudhagra** [reply]
Sorry to hear that...can you please post the outputs of the commands I listed previously? (`sudo systemctl status nginx`, `which nginx`)

And yes I understand what an SD card image is—however, security is still an issue. I have no idea what low-level system files could accidentally include sensitive information and don't want to risk that (for example `/etc/wpa_supplicant/wpa_supplicant.conf` would contain my WiFi network's SSID and password in plain text). Perhaps in the future I'll package everything in a docker container.

**Reply by MaxAltera** [reply]
Thank you for your help. Indeed, I have a problem with nginx.conf. The instructions do not say anything about this file. Found an example of this file https://gist .github.com/nishantmodak/d08aae033775cb1a0f8a Does it need to be adjusted?
![photo_2023-04-14_09-47-23](https://user-images.githubusercontent.com/86003419/231967540-fa3426e8-40f2-4b30-ab16-e9dd4bdf198f.jpg)

**Reply by rdudhagra** [reply]
Ah the commands I listed earlier are two separate commands. I got the output of the first—can you please run `which nginx` and post the output?

Still not sure why `nginx.conf` doesn't exist. From your earlier message it seems like `/etc/nginx` is empty? Which it shouldn't be according to the instructions (there should be some conf files inside a folder called `sites-available`).

**Comment by kieraneglin** [comment]
@MaxAltera I've set mine up to not use the included software at all, instead using Klipper with Fluidd as the UI. I've made serveral PRs to Fluidd to allow it to display plotter-style GCode, etc. That would be compatible with an Orange Pi and offer MANY nice-to-haves out of the box compared to Marlin.

I'll be releasing a full build guide once I'm done, but please let me know if you have any questions! Klipper has been a great success for me so far

**Reply by MaxAltera** [reply]
Thank you for your reply! I had previously planned to use Gcode playlists via OctoPrint and everything works fine, but the problem was in the management of NeoPixel. I want to use one interface that combines everything: both movement and light.  
PS: To control the LED strip, I liked the WLED project.

**Reply by kieraneglin** [reply]
Fair enough! I haven't used Octoprint in a long time, but I can say that Klipper + Fluidd + Klipper LED plugin can control everything within one interface. 

Anyway, good luck setting everything up!
